# Exercise 6

## SVM & Regularization


For this homework we consider a set of observations on a number of red and white wine varieties involving their chemical properties and ranking by tasters. Wine industry shows a recent growth spurt as social drinking is on the rise. The price of wine depends on a rather abstract concept of wine appreciation by wine tasters, opinion among whom may have a high degree of variability. Pricing of wine depends on such a volatile factor to some extent. Another key factor in wine certification and quality assessment is physicochemical tests which are laboratory-based and takes into account factors like acidity, pH level, presence of sugar and other chemical properties. For the wine market, it would be of interest if human quality of tasting can be related to the chemical properties of wine so that certification and quality assessment and assurance process is more controlled.

Two datasets are available of which one dataset is on red wine and have 1599 different varieties and the other is on white wine and have 4898 varieties. All wines are produced in a particular area of Portugal. Data are collected on 12 different properties of the wines one of which is Quality, based on sensory data, and the rest are on chemical properties of the wines including density, acidity, alcohol content etc. All chemical properties of wines are continuous variables. Quality is an ordinal variable with possible ranking from 1 (worst) to 10 (best). Each variety of wine is tasted by three independent tasters and the final rank assigned is the median rank given by the tasters.

A predictive model developed on this data is expected to provide guidance to vineyards regarding quality and price expected on their produce without heavy reliance on volatility of wine tasters.

In [2]:
import pandas as pd
import numpy as np

In [3]:
data_r = pd.read_csv('https://github.com/albahnsen/PracticalMachineLearningClass/raw/master/datasets/Wine_data_red.csv')
data_w = pd.read_csv('https://github.com/albahnsen/PracticalMachineLearningClass/raw/master/datasets/Wine_data_white.csv')

In [5]:
data = data_w.assign(type = 'white')

data = data.append(data_r.assign(type = 'red'), ignore_index=True)
data.sample(5)

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
5503            8.3              0.60         0.13            2.60      0.085   
3746            7.4              0.27         0.28            1.80      0.040   
1993            7.4              0.30         0.22            5.25      0.053   
1308            9.2              0.15         0.68            1.60      0.046   
3640            6.8              0.19         0.33            4.90      0.047   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
5503                  6.0                  24.0  0.99840  3.31       0.59   
3746                 45.0                 121.0  0.99043  3.02       0.40   
1993                 33.0                 180.0  0.99260  3.13       0.45   
1308                 22.0                 130.0  0.99480  3.02       0.45   
3640                 42.0                 130.0  0.99283  3.12       0.56   

      alcohol  quality   type  
5503      9.2        6    red  
3746     11.9        5  white  
1993     11.6        6  white  
1308     10.4        6  white  
3640     11.0        6  white

# Exercise 6.1

Show the frecuency table of the quality by type of wine

In [18]:
tabla_de_frecuencia = pd.DataFrame(data['quality'].groupby(data['type']).value_counts())
tabla_de_frecuencia

quality
type  quality         
red   5            681
      6            638
      7            199
      4             53
      8             18
      3             10
white 6           2198
      5           1457
      7            880
      8            175
      4            163
      3             20
      9              5

In [23]:
#tab= pivot_table(data, values='quality', index='quality', columns=['type'],  aggfunc='count')

In [32]:
data['quality_good'] = data["quality"] > 5 

In [35]:
datared = data[data.quality > 5 ]
datared.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
13            6.6              0.16         0.40             1.5      0.044   
15            6.6              0.17         0.38             1.5      0.032   
17            6.2              0.66         0.48             1.2      0.029   
20            6.2              0.66         0.48             1.2      0.029   
21            6.4              0.31         0.38             2.9      0.038   

    free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
13                 48.0                 143.0   0.9912  3.54       0.52   
15                 28.0                 112.0   0.9914  3.25       0.55   
17                 29.0                  75.0   0.9892  3.33       0.39   
20                 29.0                  75.0   0.9892  3.33       0.39   
21                 19.0                 102.0   0.9912  3.17       0.35   

    alcohol  quality   type  quality_good  
13     12.4        7  white          True  
15     11.4        7  white          True  
17     12.8        8  white          True  
20     12.8        8  white          True  
21     11.0        7  white          True

In [36]:
datawhite = data[data.quality <6]
datawhite.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
10            8.1              0.27         0.41            1.45      0.033   
11            8.6              0.23         0.40            4.20      0.035   
12            7.9              0.18         0.37            1.20      0.040   
14            8.3              0.42         0.62           19.25      0.040   
19            6.5              0.31         0.14            7.50      0.044   

    free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
10                 11.0                  63.0   0.9908  2.99       0.56   
11                 17.0                 109.0   0.9947  3.14       0.53   
12                 16.0                  75.0   0.9920  3.18       0.63   
14                 41.0                 172.0   1.0002  2.98       0.67   
19                 34.0                 133.0   0.9955  3.22       0.50   

    alcohol  quality   type  quality_good  
10     12.0        5  white         False  
11      9.7        5  white         False  
12     10.8        5  white         False  
14      9.7        5  white         False  
19      9.5        5  white         False

In [31]:
tabla_de_frecuencia2 = pd.DataFrame(data['quality_good'].groupby(data['type']).value_counts())
data['quality_good'].value_counts()

True     4113
False    2384
Name: quality_good, dtype: int64

# SVM

# Exercise 6.2

* Standarized the features (not the quality)
* Create a binary target for each type of wine
* Create two Linear SVM's for the white and red wines, repectively.


# Exercise 6.3

Test the two SVM's using the different kernels (‘poly’, ‘rbf’, ‘sigmoid’)


# Exercise 6.4
Using the best SVM find the parameters that gives the best performance

'C': [0.1, 1, 10, 100, 1000], 'gamma': [0.01, 0.001, 0.0001]

# Exercise 6.5

Compare the results with other methods

# Regularization

# Exercise 6.6


* Train a linear regression to predict wine quality (Continous)

* Analyze the coefficients

* Evaluate the RMSE

# Exercise 6.7

* Estimate a ridge regression with alpha equals 0.1 and 1.
* Compare the coefficients with the linear regression
* Evaluate the RMSE

# Exercise 6.8

* Estimate a lasso regression with alpha equals 0.01, 0.1 and 1.
* Compare the coefficients with the linear regression
* Evaluate the RMSE

# Exercise 6.9

* Create a binary target

* Train a logistic regression to predict wine quality (binary)

* Analyze the coefficients

* Evaluate the f1score

# Exercise 6.10

* Estimate a regularized logistic regression using:
* C = 0.01, 0.1 & 1.0
* penalty = ['l1, 'l2']
* Compare the coefficients and the f1score